src/dashboard/dashboard-custom-widget/dashboard-custom-widget.ts

```diff
@@ -0,0 +1,64 @@
+import type { DDClient } from '../../client/client';
+import {
+    UiAppEventType,
+    UiAppFeatureType,
+    UiAppRequestType
+} from '../../constants';
+import { DDFeatureClient } from '../../shared/feature-client';
+import type {
+    GetDashboardCustomWidgetOptionsResponse,
+    WidgetOptionItem
+} from '../../types';
+
+const emptyConfig: GetDashboardCustomWidgetOptionsResponse = { widgets: [] };
+
+export class DDDashboardCustomWidgetClient extends DDFeatureClient {
+    constructor(client: DDClient) {
+        super(client, UiAppFeatureType.DASHBOARD_CUSTOM_WIDGET);
+
+        // initialize with an empty reponse handler
+        this.onRequest(() => emptyConfig);
+    }
+
+    /**
+     * Registers a request handler for providing custom widget items dynamically
+     */
+    onRequest(
+        requestHandler: () =>
+            | GetDashboardCustomWidgetOptionsResponse
+            | Promise<GetDashboardCustomWidgetOptionsResponse>
+    ) {
+        this.client.framePostClient.onRequest(
+            UiAppRequestType.GET_DASHBOARD_CUSTOM_WIDGET_ITEMS,
+            async (): Promise<GetDashboardCustomWidgetOptionsResponse> => {
+                await this.validateFeatureIsEnabled();
+
+                const { widgets } = await requestHandler();
+
+                return { widgets };
+            }
+        );
+
+        // return an unsubscribe hook
+        return () => {
+            this.onRequest(() => emptyConfig);
+        };
+    }
+
+    async updateOptions(
+        optionsToAdd: WidgetOptionItem[],
+        optionsNamesToRemove: string[] = []
+    ) {
```

Comment:
I imagine the other option more redundant because I'd still need to maintain and pass options that didn't change again and again instead of passing only what changed, but I agree it might be a good idea to get more feedback. @hydrosquall what do you think?

Reply:
I think having a way to updating existing options that doesn't involve making 2 calls would be useful. For example in the [demo app](https://github.com/DataDog/apps-starter-kit/compare/rami/dynamic_options)

- Updating the members of APP to have different versions based on some polling interval (e.g. every minute)
- Updating the options from some static config (e.g. checking which of the cheeses in the dropdown are currently in stock, and removing the ones that are out of stock)

Currently, it seems like someone would need to make 2 calls: 1 to remove the old option , and 1 to add the new options. I think a single `updateOptions( newOptions)` call might be easier to reason about. The downside is that they'd need a copy of the old options state. I was assuming they'd have access to this via the callback in the `DASHBOARD_CUSTOM_WIDGET_OPTIONS_CHANGE` event rather than needing to keep track of some additional local state themselves.